pkg/controller/atlasdeployment/search_nodes.go

```diff
@@ -0,0 +1,312 @@
+		return s.delete()
+	default:
+		// akoEmpty && atlasEmpty
+		// If no search nodes are in AKO and no nodes in Atlas - unmanage.
+		return s.unmanage()
+	}
+}
+
+// handleUpserting handles creating or updating search nodes. The following state is handled here:
+//
+//	Conditions:
+//	  Status:                False
+//	  Type:                  SearchNodesReady
+//	  Reason:                SearchNodesCreating | SearchNodesUpdating
+//
+// The following transitions can happen here:
+// - terminate: when an error occurred or when the operation has been aborted.
+// - idle: when search nodes are marked as IDLE in Atlas.
+func (s *searchNodeController) handleUpserting(state workflow.ConditionReason) workflow.Result {
+	if len(s.deployment.Spec.DeploymentSpec.SearchNodes) == 0 {
+		return s.terminate(workflow.ErrorSearchNodesOperationAborted, errors.New("aborting update/create: no search nodes specified"))
+	}
+
+	atlasNodes, found, err := s.getAtlasSearchDeployment()
+	switch {
+	case err != nil:
+		return s.terminate(workflow.ErrorSearchNodesNotUpsertedInAtlas, err)
+	case !found:
+		return s.terminate(workflow.ErrorSearchNodesNotUpsertedInAtlas, errors.New("no search nodes found in Atlas"))
+	case atlasNodes.GetStateName() != "IDLE":
+		return s.progress(
+			state,
+			fmt.Sprintf("search nodes are not ready yet, Atlas state: %q", atlasNodes.GetStateName()),
+			"waiting for search nodes to become ready",
+		)
+	default:
+		return s.idle()
+	}
+}
+
+// handleCreating handles creating search nodes. The following state is handled here:
+//
+//	Conditions:
+//	  Status:                False
+//	  Type:                  SearchNodesReady
+//	  Reason:                SearchNodesDeleting
+//
+// The following transitions can happen here:
+// - terminate: when an error occurred or when the operation has been aborted.
+// - unmanage: when there are no search nodes anymore in Atlas.
+func (s *searchNodeController) handleDeleting() workflow.Result {
+	if len(s.deployment.Spec.DeploymentSpec.SearchNodes) > 0 {
+		return s.terminate(workflow.ErrorSearchNodesOperationAborted, errors.New("aborting deletion: search nodes are specified"))
+	}
+
+	atlasNodes, found, err := s.getAtlasSearchDeployment()
+	if err != nil {
+		// transition back to pending, something went wrong.
+		return s.terminate(workflow.ErrorSearchNodesNotDeletedInAtlas, err)
+	}
+
+	switch {
+	case err != nil:
+		return s.terminate(workflow.ErrorSearchNodesNotUpsertedInAtlas, err)
+	case found:
+		return s.progress(
+			workflow.SearchNodesDeleting,
+			fmt.Sprintf("search nodes are being deleted, Atlas state: %q", atlasNodes.GetStateName()),
+			"deleting search nodes",
+		)
+	default:
+		return s.unmanage()
+	}
+}
+
+// create executes the actual creation of search nodes in Atlas and transitions to the following states:
+// - creating: after search nodes have been created in Atlas
+// - terminated: when an error occurred.
+func (s *searchNodeController) create() workflow.Result {
```

Comment:
Nit: this almost matches a translation layer function on top of the API call. But I would probably not include workflow.Result in such a layer.

Reply:
this is a state machine transition method so it must return a `workflow.Result`

Reply:
Yes, I think when we add the translation layer it should wrap it here. The code will be a bit silly, just a call to the translation layer and handling the return into a proper `workflow.Result`. Note that Contract Tests should NOT be aware of any workflow.Result, it is not its concern.

Reply:
I think that's fine but out of scope for a `create() workflow.Result` state machine transition method 👍